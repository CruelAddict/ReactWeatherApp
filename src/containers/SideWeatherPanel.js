import React from 'react'
import WeatherPanel from './WeatherPanel'
import SideWeatherPanelComponent from '../components/SideWeatherPanelComponent'
import {connect} from 'react-redux';
import Actions from "../modules/actions";
import {NotificationManager} from 'react-notifications';

const mapStateToProps = state => ({
    ...state
});

const mapDispatchToProps = dispatch => ({
    deleteFavorite: id => dispatch(Actions.deleteFavorite(id)),
    setFavoriteWeather: (favorite, id) => dispatch(Actions.setFavoriteWeather(favorite, id)),
    markAsPosted: id => dispatch(Actions.markAsPosted(id))
});


class SideWeatherPanel extends WeatherPanel {

    constructor() {
        super();
        this.handleFavoriteRemoval = this.handleFavoriteRemoval.bind(this);
        this.setWeather = this.setWeather.bind(this);
        this.postFavourite = this.postFavourite.bind(this)
    }

    componentDidMount() {
        this.setWeather();
        this.setState(
            this.props.favorites.items.filter(value => (value.id === this.props.favoriteId))[0].weatherObj
        );
    }

    async postFavourite(newState) {
        await   this.props.markAsPosted(this.props.favoriteId);
        console.log('posting favorite');
        fetch('http://127.0.0.1:3000/favourites/', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json;charset=utf-8'
            },
            body: JSON.stringify({
                name : newState.city
            })
        }).then(response => {
            console.log(response);
        }, response => {
            console.log('Error!');
            console.log(response);
            NotificationManager.error('К сожалению, после перезагрузки страницы изменения будут потеряны', 'Не удалось сохранить избранный город');
        });
    }

    componentDidUpdate(prevProps) {
        let favoriteObj = this.props.favorites.items.filter(value => (value.id === this.props.favoriteId))[0];

        let oldData = prevProps.favorites.items.filter(value => (value.id === this.props.favoriteId))[0].weatherObj;
        let newData = favoriteObj.weatherObj;
        if(!favoriteObj.pushedToServer && !newData.isFailed) {
            oldData === undefined && newData !== undefined && this.postFavourite(newData);
        }
        oldData !== newData &&  this.setState(newData);
    }

    setWeather() {
        this.props.setFavoriteWeather(this, this.props.favoriteId);
    }

    getWeather() {
        return new Promise( (resolve, reject) => {
                let url = this.weatherByNameUrl();
                this.execWeatherRequest(url).then(weatherObj => (resolve(weatherObj)),
                    errorMsg => (reject(errorMsg)));
            }
        );
    }

    handleFavoriteRemoval() {
        console.log(`deleting ${this.state.city}`);
        fetch('http://127.0.0.1:3000/favourites/', {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json;charset=utf-8'
            },
            body: JSON.stringify({
                name : this.state.city
            })
        }).then(response => {
            !response.ok && NotificationManager.error('Изменения могут быть не сохранены.', 'Произошла ошибка')
        }, response => {
            NotificationManager.error('Изменения могут быть не сохранены.', 'Произошла ошибка')
        });
        this.props.deleteFavorite(this.props.favoriteId);
    }

    render() {
        return (
                <SideWeatherPanelComponent
                    state={this.state}
                    handleFavoriteRemoval={this.handleFavoriteRemoval}
                    setWeather={this.setWeather}
                />
        )
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(SideWeatherPanel);
