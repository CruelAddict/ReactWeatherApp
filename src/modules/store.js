import {createStore, applyMiddleware} from 'redux';
import thunk from 'redux-thunk';
import reducer from './rootReducer'
import store from "../index";

const savedItems = localStorage.getItem('favorites') ? JSON.parse(localStorage.getItem('favorites')).map(value => {
        return {
            ...value,
            isLoading: true
        };
    }) : [];

const initialState = {
    favorites: {
        items: savedItems
    },
    geo: {
        permissionGranted: null,
        coordinates: null
    }
};

export default () => {
    return createStore(
        reducer,
        initialState,
        applyMiddleware(thunk)
    )
}
