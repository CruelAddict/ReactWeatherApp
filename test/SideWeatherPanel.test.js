import React from 'react';
import {shallow, render, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as Enzyme from "enzyme";
import {act} from 'react-dom/test-utils';
import sinon from 'sinon';
const MockXMLHttpRequest = require('mock-xmlhttprequest');

import store from '../src/modules/store'
import SideWeatherPanel from '../src/containers/SideWeatherPanel';
import Actions from "../src/modules/actions";

var panel;

window.fetch = sinon.stub();

const imgData = 'fakeData';

window.URL.createObjectURL = sinon.stub().returns('fakeURL');

const blobStub = () => Promise.resolve();

const mockApiResponse = (body = {}) => {
    return {
        status: 200,
        headers: { 'Content-type': 'application/json' },
        body: JSON.stringify(body),
        blob: blobStub
    };
};

window.fetch.returns(Promise.resolve(mockApiResponse(imgData)));


Enzyme.configure({adapter: new Adapter()});

const server = MockXMLHttpRequest.newServer({
    get: ['https://api.openweathermap.org/data/2.5/weather?q=London&appid=bf3565940c52aaa3383c1dbc23799bb1', {
        // status: 200 is the default
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({
            "coord": {"lon": -0.13, "lat": 51.51},
            "weather": [{"id": 803, "main": "Clouds", "description": "broken clouds", "icon": "04d"}],
            "base": "stations",
            "main": {"temp": 280.6, "pressure": 1011, "humidity": 81, "temp_min": 279.82, "temp_max": 281.48},
            "visibility": 10000,
            "wind": {"speed": 2.6, "deg": 360},
            "clouds": {"all": 75},
            "dt": 1574004909,
            "sys": {"type": 1, "id": 1412, "country": "GB", "sunrise": 1573975285, "sunset": 1574006980},
            "timezone": 0,
            "id": 2643743,
            "name": "London",
            "cod": 200
        }),
    }],
}).install();

server.get('https://api.openweathermap.org/data/2.5/weather?q=CityDoesNotExist&appid=bf3565940c52aaa3383c1dbc23799bb1', {
    status: 404,
    headers: {'Content-Type': 'application/json'},
    body: 'city not found'
});


server.get('https://api.openweathermap.org/data/2.5/weather?q=ServerBroke&appid=bf3565940c52aaa3383c1dbc23799bb1', {
    status: 500,
    headers: {'Content-Type': 'application/json'},
    body: ''
});


describe('SideWeatherPanel', () => {

    it('Should display data when loaded', () => {
        store.dispatch(Actions.createFavorite("London")).then( () => {
            panel = mount(
                <SideWeatherPanel
                    favoriteId={0}
                    city={"London"}
                    store={store}
                />
            );
        });

        return new Promise(resolve => {
            let waitUntilLoaded = setInterval(() => {
                panel.find('SideWeatherPanel').instance().state.isLoading === false ? resolve(waitUntilLoaded) : null;
            }, 250)
        }).then((waitUntilLoaded) => {
            clearInterval(waitUntilLoaded);
            panel.update();
            expect(panel).toMatchSnapshot();
        });
    });

    it('Should display error message if city does not respond', () => {
        store.dispatch(Actions.createFavorite("ServerBroke")).then( () => {
            panel = mount(
                <SideWeatherPanel
                    favoriteId={1}
                    city={"ServerBroke"}
                    store={store}
                />
            );
        });

        return new Promise(resolve => {
            let waitUntilLoaded = setInterval(() => {
                console.log(panel.find('SideWeatherPanel').instance().state);
                panel.find('SideWeatherPanel').instance().state.isLoading === false ? resolve(waitUntilLoaded) : null;
            }, 250)
        }).then((waitUntilLoaded) => {
            clearInterval(waitUntilLoaded);
            panel.update();
            expect(panel).toMatchSnapshot();
        });
    });

    it('Should display error message if city not found', () => {
        store.dispatch(Actions.createFavorite("CityDoesNotExist")).then( () => {
            panel = mount(
                <SideWeatherPanel
                    favoriteId={2}
                    city={"CityDoesNotExist"}
                    store={store}
                />
            );
        });

        return new Promise(resolve => {
            let waitUntilLoaded = setInterval(() => {
                console.log(panel.find('SideWeatherPanel').instance().state);
                panel.find('SideWeatherPanel').instance().state.isLoading === false ? resolve(waitUntilLoaded) : null;
            }, 250)
        }).then((waitUntilLoaded) => {
            clearInterval(waitUntilLoaded);
            panel.update();
            expect(panel).toMatchSnapshot();
        });
    });


    it('Should initially be loading', () => {
        window.fetch = sinon.stub().returns(setTimeout(()=> Promise.resolve(), 15000));
        store.dispatch(Actions.createFavorite("London")).then( () => {
            panel = mount(
                <SideWeatherPanel
                    favoriteId={3}
                    city={"London"}
                    store={store}
                />
            );
        }).then(() => {
            panel.update();
            expect(panel).toMatchSnapshot();
        });
    });

});
