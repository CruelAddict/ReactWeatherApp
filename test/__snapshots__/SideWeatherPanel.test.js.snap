// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideWeatherPanel Should display data when loaded 1`] = `
<Connect(SideWeatherPanel)
  city="London"
  favoriteId={0}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <SideWeatherPanel
    city="London"
    deleteFavorite={[Function]}
    favoriteId={0}
    favorites={
      Object {
        "items": Array [
          Object {
            "id": 0,
            "name": "London",
            "pushedToServer": true,
            "weatherObj": Object {
              "city": "London",
              "coordinates": Object {
                "latitude": 51.51,
                "longitude": -0.13,
              },
              "isFailed": false,
              "isLoading": false,
              "weather_data": Object {
                "cloudiness": "broken clouds",
                "humidity": 81,
                "icon": "fakeURL",
                "pressure": 1011,
                "temperature": 7.4500000000000455,
                "wind": 2.6,
              },
            },
          },
        ],
      }
    }
    geo={
      Object {
        "coordinates": null,
        "permissionGranted": null,
      }
    }
    markAsPosted={[Function]}
    setFavoriteWeather={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <_default
      handleFavoriteRemoval={[Function]}
      setWeather={[Function]}
      state={
        Object {
          "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
          "city": "London",
          "coordinates": Object {
            "latitude": 51.51,
            "longitude": -0.13,
          },
          "isFailed": false,
          "isLoading": false,
          "weather_data": Object {
            "cloudiness": "broken clouds",
            "humidity": 81,
            "icon": "fakeURL",
            "pressure": 1011,
            "temperature": 7.4500000000000455,
            "wind": 2.6,
          },
        }
      }
    >
      <div
        className="side-weather-panel-container"
      >
        <div
          className="side-weather-panel-main-data"
        >
          <div
            className="side-city-name-container"
          >
            <h2>
              London
            </h2>
          </div>
          <div
            className="side-temperature-container"
          >
            <span>
              7
              °С
            </span>
          </div>
          <div
            className="side-icon-cell"
          >
            <div
              className="side-icon-container"
            >
              <img
                alt="weather icon"
                src="fakeURL"
              />
            </div>
          </div>
          <div
            className="delete-favorite-button-container"
          >
            <button
              className="delete-favorite favorites-action-button"
              onClick={[Function]}
            >
              ×
            </button>
          </div>
        </div>
        <div
          className="side-weather-items"
        >
          <_default
            state={
              Object {
                "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
                "city": "London",
                "coordinates": Object {
                  "latitude": 51.51,
                  "longitude": -0.13,
                },
                "isFailed": false,
                "isLoading": false,
                "weather_data": Object {
                  "cloudiness": "broken clouds",
                  "humidity": 81,
                  "icon": "fakeURL",
                  "pressure": 1011,
                  "temperature": 7.4500000000000455,
                  "wind": 2.6,
                },
              }
            }
          >
            <div
              className="weather-value-list"
            >
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Ветер
                </span>
                <span
                  className="weather-value-value"
                >
                  2.6
                   m/s
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Облачность
                </span>
                <span
                  className="weather-value-value"
                >
                  Broken Clouds
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Давление
                </span>
                <span
                  className="weather-value-value"
                >
                  1011
                   hpa
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Влажность
                </span>
                <span
                  className="weather-value-value"
                >
                  81
                   %
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Координаты
                </span>
                <span
                  className="weather-value-value"
                >
                  [
                  -0.13
                  , 
                  51.51
                  ]
                </span>
              </div>
            </div>
          </_default>
        </div>
      </div>
    </_default>
  </SideWeatherPanel>
</Connect(SideWeatherPanel)>
`;

exports[`SideWeatherPanel Should display error message if city does not respond 1`] = `
<Connect(SideWeatherPanel)
  city="ServerBroke"
  favoriteId={1}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <SideWeatherPanel
    city="ServerBroke"
    deleteFavorite={[Function]}
    favoriteId={1}
    favorites={
      Object {
        "items": Array [
          Object {
            "id": 0,
            "name": "London",
            "pushedToServer": true,
            "weatherObj": Object {
              "city": "London",
              "coordinates": Object {
                "latitude": 51.51,
                "longitude": -0.13,
              },
              "isFailed": false,
              "isLoading": false,
              "weather_data": Object {
                "cloudiness": "broken clouds",
                "humidity": 81,
                "icon": "fakeURL",
                "pressure": 1011,
                "temperature": 7.4500000000000455,
                "wind": 2.6,
              },
            },
          },
          Object {
            "id": 1,
            "name": "ServerBroke",
            "pushedToServer": false,
            "weatherObj": Object {
              "errorMessage": "Сервер не отвечает",
              "isFailed": true,
              "isLoading": false,
            },
          },
        ],
      }
    }
    geo={
      Object {
        "coordinates": null,
        "permissionGranted": null,
      }
    }
    markAsPosted={[Function]}
    setFavoriteWeather={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <_default
      handleFavoriteRemoval={[Function]}
      setWeather={[Function]}
      state={
        Object {
          "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
          "errorMessage": "Сервер не отвечает",
          "isFailed": true,
          "isLoading": false,
        }
      }
    >
      <div
        className="side-weather-panel-container"
      >
        <div
          className="side-weather-panel-main-data"
        >
          <div
            className="side-city-name-container"
          >
            <h2 />
          </div>
          <div
            className="side-temperature-container"
          />
          <div
            className="side-icon-cell"
          >
            <div />
          </div>
          <div
            className="delete-favorite-button-container"
          >
            <button
              className="delete-favorite favorites-action-button"
              onClick={[Function]}
            >
              ×
            </button>
          </div>
        </div>
        <div
          className="loading-side"
        >
          <h4
            className="loading-title-side side-error-message"
          >
            Сервер не отвечает
          </h4>
          <div
            className="refresh-button-container"
          >
            <button
              className="refresh-button"
              onClick={[Function]}
            >
              Попробовать снова
            </button>
          </div>
        </div>
      </div>
    </_default>
  </SideWeatherPanel>
</Connect(SideWeatherPanel)>
`;

exports[`SideWeatherPanel Should display error message if city not found 1`] = `
<Connect(SideWeatherPanel)
  city="CityDoesNotExist"
  favoriteId={2}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <SideWeatherPanel
    city="CityDoesNotExist"
    deleteFavorite={[Function]}
    favoriteId={2}
    favorites={
      Object {
        "items": Array [
          Object {
            "id": 0,
            "name": "London",
            "pushedToServer": true,
            "weatherObj": Object {
              "city": "London",
              "coordinates": Object {
                "latitude": 51.51,
                "longitude": -0.13,
              },
              "isFailed": false,
              "isLoading": false,
              "weather_data": Object {
                "cloudiness": "broken clouds",
                "humidity": 81,
                "icon": "fakeURL",
                "pressure": 1011,
                "temperature": 7.4500000000000455,
                "wind": 2.6,
              },
            },
          },
          Object {
            "id": 1,
            "name": "ServerBroke",
            "pushedToServer": false,
            "weatherObj": Object {
              "errorMessage": "Сервер не отвечает",
              "isFailed": true,
              "isLoading": false,
            },
          },
          Object {
            "id": 2,
            "name": "CityDoesNotExist",
            "pushedToServer": false,
            "weatherObj": Object {
              "errorMessage": "Город не найден",
              "isFailed": true,
              "isLoading": false,
            },
          },
        ],
      }
    }
    geo={
      Object {
        "coordinates": null,
        "permissionGranted": null,
      }
    }
    markAsPosted={[Function]}
    setFavoriteWeather={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <_default
      handleFavoriteRemoval={[Function]}
      setWeather={[Function]}
      state={
        Object {
          "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
          "errorMessage": "Город не найден",
          "isFailed": true,
          "isLoading": false,
        }
      }
    >
      <div
        className="side-weather-panel-container city-not-found"
      >
        <div
          className="side-weather-panel-main-data"
        >
          <div
            className="side-city-name-container"
          >
            <h2 />
          </div>
          <div
            className="side-temperature-container"
          />
          <div
            className="side-icon-cell"
          >
            <div />
          </div>
          <div
            className="delete-favorite-button-container"
          >
            <button
              className="delete-favorite favorites-action-button"
              onClick={[Function]}
            >
              ×
            </button>
          </div>
        </div>
        <div
          className="loading-side"
        >
          <h4
            className="loading-title-side side-error-message"
          >
            Город не найден
          </h4>
        </div>
      </div>
    </_default>
  </SideWeatherPanel>
</Connect(SideWeatherPanel)>
`;

exports[`SideWeatherPanel Should initially be loading 1`] = `
<Connect(SideWeatherPanel)
  city="London"
  favoriteId={3}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <SideWeatherPanel
    city="London"
    deleteFavorite={[Function]}
    favoriteId={3}
    favorites={
      Object {
        "items": Array [
          Object {
            "id": 0,
            "name": "London",
            "pushedToServer": true,
            "weatherObj": Object {
              "city": "London",
              "coordinates": Object {
                "latitude": 51.51,
                "longitude": -0.13,
              },
              "isFailed": false,
              "isLoading": false,
              "weather_data": Object {
                "cloudiness": "broken clouds",
                "humidity": 81,
                "icon": "fakeURL",
                "pressure": 1011,
                "temperature": 7.4500000000000455,
                "wind": 2.6,
              },
            },
          },
          Object {
            "id": 1,
            "name": "ServerBroke",
            "pushedToServer": false,
            "weatherObj": Object {
              "errorMessage": "Сервер не отвечает",
              "isFailed": true,
              "isLoading": false,
            },
          },
          Object {
            "id": 2,
            "name": "CityDoesNotExist",
            "pushedToServer": false,
            "weatherObj": Object {
              "errorMessage": "Город не найден",
              "isFailed": true,
              "isLoading": false,
            },
          },
          Object {
            "id": 3,
            "name": "London",
            "pushedToServer": false,
          },
        ],
      }
    }
    geo={
      Object {
        "coordinates": null,
        "permissionGranted": null,
      }
    }
    markAsPosted={[Function]}
    setFavoriteWeather={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <_default
      handleFavoriteRemoval={[Function]}
      setWeather={[Function]}
      state={
        Object {
          "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
          "isLoading": true,
        }
      }
    >
      <div
        className="side-weather-panel-container"
      >
        <div
          className="side-weather-panel-main-data"
        >
          <div
            className="side-city-name-container"
          >
            <h2 />
          </div>
          <div
            className="side-temperature-container"
          />
          <div
            className="side-icon-cell"
          >
            <div />
          </div>
          <div
            className="delete-favorite-button-container"
          >
            <button
              className="delete-favorite favorites-action-button"
              onClick={[Function]}
            >
              ×
            </button>
          </div>
        </div>
        <div
          className="loading-side"
        >
          <h4
            className="loading-title-side"
          >
            Подождите, данные загружаются
          </h4>
          <div
            className="spinner-container"
          >
            <i
              className="fa fa-refresh fa-spin fa-4x"
            />
          </div>
        </div>
      </div>
    </_default>
  </SideWeatherPanel>
</Connect(SideWeatherPanel)>
`;
